### ISSUE TITLE
automated language selection

### ISSUE THREAD

**Issue by HexeBianca** [IMAGINARY/imaginary-web#52]
When opening www.imaginary.org it would be great if the language of the browser can be automatically used to switch to the corresponding language (if available).
According to Andreas, this used to work for the old website.
Can this be done togehter with the launch of the French version?

**Comment by elondaits** [comment]
Yes. It's currently disabled. I could enable it just now but I'll think of the right language criteria judiciously (Drupal allows for a complex criteria combining the browser, URLs and user settings) and test it to see it works right.

**Comment by HexeBianca** [comment]
great!

**Comment by christophknoth** [comment]
From what i heard that is a very tricky subject. A couple of big sites do this quite well, but I have also visited sites where I was lost in translation. When I am visiting turkey will it send me to the language of my IP (Turkish) or my browser (English) or my operating system (German)?

**Comment by elondaits** [comment]
I found that many sites use Geolocation to send you to the language of your IP... but this is WRONG. I'm the perfect example because here in Europe, browsing with my computer, I get german and french versions of webpages instead of the english / spanish which I've configured in my browser as my preferred languages.

Browsers allow for the configuration of an ordered list of preferred languages and you can negotiate with that (Drupal does this)... even for someone that never bothered to configure the default should be the language of the browser, which is a decent default (e.g. spanish speaking people normally download the spanish version of the browser).

Having said all this I found some problems in testing the language selection through browser preferences in Imaginary. Basically two problems so far:

1 - (The hard problem) What Drupal does when changing languages based on the browser configuration is to show the preferred language in the "default" site pages (the ones that are not prefixed by the language... e.g. www.imaginary.org would be the default, www.imaginary.org/de would be the german page). So if you have a spanish browser you'll see spanish in www.imaginary.org/ instead of begin redirected to www.imaginary.org/es .

2 - (The not as hard problem) The language selector ignores the language hiding module... so people with a browser configured for a hidden language get an "access denied" page instead of content.

I could make my own module and bypass both problems... but there's another one:

3 - (The REAL problem) The "default" pages are the ones in english... how can I tell if I should redirect the user to his preferred language or if he actually wants to navigate the english pages? 

I'll research the issue a bit more to see if other sites have a good way of solving this. Considering how Google solves it (poorly) I'm not sure I'll find a solution.

**Comment by HexeBianca** [comment]
I don't see the Problem in 1. Can you explain, Eric?
2. would have to be changed.
3. Can the user still switch languages via the language selector?

**Comment by elondaits** [comment]
1 - I'll explain hopefully better:

- **Right now**:
  - www.imaginary.org: English version
  - www.imaginary.org/es: Spanish version
  - www.imaginary.org/de: German version
- **With Drupal's browser-based translation enabled and a spanish [configured] browser**:
  - www.imaginary.org: Spanish version
  - www.imaginary.org/es: Spanish version
  - www.imaginary.org/de: German version

The "default" site uses the detected language and there's no way to access the english version. This is why this configuration is not useful for us, besides the problem with the "hidden language" module.

3 - The only thing the language selector does is redirect the user to another URL. Suppose I create a module that instead of changing the default URL of the site to the browser's language redirects to the right language... so you access www.imaginary.org and I check the browser, see it's configured for spanish and redirect you to www.imaginary.org/es ... so far OK... but imagine you actually want the english version of the site... so you select english in the language selector and get redirected back to www.imaginary.org... where I'd check your browser and redirect you back to www.imaginary.org/es. A solution that takes into account "the user explicitly chose english" is what some sites use, but it's not easy to do right and still leads to some frustration.

**Comment by HexeBianca** [comment]
Thank you, Eric!
For 1. could we introduce www.imaginary.org/en additionally, which can be accessed via the language selector?
Would this may be solve 3. as well?
The automatic language selection should only be active for www.imaginary.org. If a user chooses a different langugae via the language selector, no automatism should interfere.

**Comment by elondaits** [comment]
That *could* be an option, perhaps... but I'm not sure Drupal supports
it... I'll have to check. It could also be problematic with Google, who
penalizes having duplicate content... we'd have to force Google to see
english at the default address and ban it from /en, since we want to keep
the score it gave to the current URLs. I'll have to check that too.

**Comment by christophknoth** [comment]
I think we should stick to english without using /en/ as all links point there.

I would really like to postpone this whole issue. This is definitely not high priority and not easy to solve.

**Comment by christophknoth** [comment]
Alternatively we could work on overlay that says: It seems like your main language is XYZ maybe you want to check out our site in this language and then a Yes and No button.

But I am not sure if it is super important and helpful. If I come to the site I will either come from google which will already list results in the language I searched or I come via a link where the link had been set for a certain language. It might be weird if we forward people from there.
